cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java

```diff
@@ -64,12 +64,12 @@ public void process(ApplicationDeployable input) throws Exception {
         // note: we can deploy module or create module from Dataset class
         // note: it seems dangerous to instantiate dataset module here, but this will be fine when we move deploy into
         //       isolated user's environment (e.g. separate yarn container)
-        Id.DatasetModule moduleId = Id.DatasetModule.from(input.getId().getNamespaceId(), moduleName);
+        Id.DatasetModule moduleId = Id.DatasetModule.from(input.getId().getNamespace(), moduleName);
         if (DatasetModule.class.isAssignableFrom(clazz)) {
           datasetFramework.addModule(moduleId, (DatasetModule) clazz.newInstance());
         } else if (Dataset.class.isAssignableFrom(clazz)) {
           // checking if type is in already
-          if (!datasetFramework.hasType(clazz.getName())) {
+          if (!datasetFramework.hasSystemType(clazz.getName())) {
```

Comment:
Yeah, I wanted to get feedback on this. The current logic is that we will not upgrade if a module exists, irrespective of whether its a user module. Ideally we should allow upgrading user modules by default, but I understand there are some limitations regarding this in CDAP currently.

I'm unsure which way to go here amongst the following 3 options:
1. Allow upgrade no matter what - even if it is a module that exists in the system namespace, we create a copy of it in another namespace and allow users to use it. 
2. Allow upgrade only for user modules - the logic that is used in this PR.
3. Do not allow upgrade at all - the current logic.

I guess all three are possible from the namespaces point of view, however, I understand that other things in datasets can break with upgrades.